cmd/gas/main.go

```diff
@@ -0,0 +1,232 @@
+	flagRulesInclude = flag.String("include", "", "Comma separated list of rules IDs to include. (see rule list)")
+
+	// rules to explicitly exclude
+	flagRulesExclude = flag.String("exclude", "", "Comma separated list of rules IDs to exclude. (see rule list)")
+
+	// log to file or stderr
+	flagLogfile = flag.String("log", "", "Log messages to file rather than stderr")
+
+	logger *log.Logger
+)
+
+// #nosec
+func usage() {
+
+	fmt.Fprintln(os.Stderr, usageText)
+	fmt.Fprint(os.Stderr, "OPTIONS:\n\n")
+	flag.PrintDefaults()
+	fmt.Fprint(os.Stderr, "\n\nRULES:\n\n")
+
+	// sorted rule list for eas of reading
+	rl := rules.Generate()
+	keys := make([]string, 0, len(rl))
+	for key := range rl {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		v := rl[k]
+		fmt.Fprintf(os.Stderr, "\t%s: %s\n", k, v.Description)
+	}
+	fmt.Fprint(os.Stderr, "\n")
+}
+
+func loadConfig(configFile string) (gas.Config, error) {
+	config := gas.NewConfig()
+	if configFile != "" {
+		file, err := os.Open(configFile)
+		if err != nil {
+			return nil, err
+		}
+		defer file.Close()
+		if _, err := config.ReadFrom(file); err != nil {
+			return nil, err
+		}
+	}
+	if *flagIgnoreNoSec {
+		config.SetGlobal("nosec", "true")
+	}
+	return config, nil
+}
+
+func loadRules(include, exclude string) rules.RuleList {
+	filters := make([]rules.RuleFilter, 0)
+	if include != "" {
+		including := strings.Split(include, ",")
+		filters = append(filters, rules.NewRuleFilter(false, including...))
+	}
+
+	if exclude != "" {
+		excluding := strings.Split(exclude, ",")
+		filters = append(filters, rules.NewRuleFilter(true, excluding...))
+	}
+	return rules.Generate(filters...)
+}
+
+func saveOutput(filename, format string, issues []*gas.Issue, metrics *gas.Metrics) error {
+	if filename != "" {
+		outfile, err := os.Create(filename)
+		if err != nil {
+			return err
+		}
+		defer outfile.Close()
+		output.CreateReport(outfile, format, issues, metrics)
+	} else {
+		output.CreateReport(os.Stdout, format, issues, metrics)
+	}
+	return nil
+}
+
+func main() {
+
+	// Setup usage description
+	flag.Usage = usage
+
+	// Parse command line arguments
+	flag.Parse()
+
+	// Ensure at least one file was specified
+	if flag.NArg() == 0 {
+		fmt.Fprintf(os.Stderr, "\nError: FILE [FILE...] or './...' expected\n")
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	// Setup logging
+	logWriter := os.Stderr
+	if *flagLogfile != "" {
+		var e error
+		logWriter, e = os.Create(*flagLogfile)
+		if e != nil {
+			flag.Usage()
+			log.Fatal(e)
+		}
+	}
+	logger = log.New(logWriter, "[gas] ", log.LstdFlags)
+
+	// Load config
+	config, err := loadConfig(*flagConfig)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	// Load enabled rule definitions
+	ruleDefinitions := loadRules(*flagRulesInclude, *flagRulesExclude)
+
+	// Create the analyzer
+	analyzer := gas.NewAnalyzer(config, logger)
+	analyzer.LoadRules(ruleDefinitions.Builders()...)
+
+	vendor := regexp.MustCompile(`[\\/]vendor([\\/]|$)`)
```

Comment:
Maybe is worth make it this configurable. I think we can get a list of folders which could be potentially excluded. I would add **vendor** folder by defualt to this list.

Reply:
The file inclusion / exclusion logic in the past has been a major source of bugs and confusion. The aim here was to basically make gas work as simply as possible. 

However I agree  it would make sense to be able to configure package directories in the recursive case. Do you think we should fix this as a separate PR or try to sneak it in with change?

Reply:
In my opinion the vendor folder exclusion is enough. A more complex logic can be added after the merge.